### ISSUE TITLE
Custom Shell.TitleView not positioned correctly on iOS

### ISSUE THREAD

**Issue by ewerspej** [dotnet/maui#14558]
### Description

The content of a custom Shell.TitleView is positioned incorrectly. Whatever I do, I cannot center the content, such as a label correctly.

I have the following code where I use a custom font, which works perfectly fine on Android, but on iOS it doesn't:

```xml
<Shell.TitleView>
  <HorizontalStackLayout VerticalOptions="Fill">
    <Label
      Text="Welcome to MAUI"
      FontFamily="Strande2"
      TextColor="White"
      VerticalTextAlignment="Center"
      VerticalOptions="CenterAndExpand"
      HeightRequest="50"
      FontSize="Medium" />
  </HorizontalStackLayout>
</Shell.TitleView>
```

The content is supposed to be vertically centered, so that the title label is displayed normally.

However, on iOS it currently looks like this:

<img src="https://github.com/ewerspej/maui-samples/blob/main/issues/titleview_ios/ios_bug_titleview.PNG?raw=true" width="500" />

On Android it looks fine:

<img src="https://github.com/ewerspej/maui-samples/blob/main/issues/titleview_ios/androind_titleview.PNG?raw=true" width="500" />

### Steps to Reproduce

1. Create new MAUI project
2. On MainPage, set `<Shell.TitleView>` with a `VerticalStackLayout` and a `Label` inside
3. Set everything like above in the description
4. Observe the incorrect behavior on iOS

### Link to public reproduction project repository

https://github.com/ewerspej/maui-samples

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 16 (possibly earlier versions, too)

### Did you find any workaround?

No

### Relevant log output

```shell
n/a
```

**Comment by kofanov** [comment]
This issue appears on new iOS devices (without physical Home button). Here is my workaround:

Create custom TitleView class:
`public class MyTitleView : ContentView {}`

Add this method to MyTitleView:
```
#if IOS
    protected override Size ArrangeOverride(Rect bounds)
    {
        if (bounds.Height == 0)
            bounds.Height = 44;

        return base.ArrangeOverride(bounds);
    }
    #endif
```

Set TitleView on OnApplyTemplate method of ContentPage:
```
protected override void OnApplyTemplate()
    {
        base.OnApplyTemplate();
        Shell.SetTitleView(this, new MyTitleView());
    }
```

**Comment by codingL3gend** [comment]
@kofanov  that worked partially for me but weird behavior seeing the title view disappear when switching to a different tab and coming back. have you noticed the same? testing on iPhone 14 iOS 16.4

**Comment by kofanov** [comment]
@codingL3gend Yes, I have encountered this problem. There are a lot of problems and "surprises" in MAUI in general. That's why I assign TitleView in OnApplyTemplate method of ContentPage (assign again and again in every single ContentPage!) and MyTitleView looks like:

```
<ContentView.Resources>
        <Style TargetType="controls:MyTitleView">
            <Setter Property="Title" Value="{Binding Title, Source={RelativeSource AncestorType={Type ContentPage}}}" />
    </ContentView.Resources>
...
<Label Text="{Binding Title}" />
```

I spent a lot of time and conducted a lot of experiments to find it out. Perhaps the Title property is redundant here, but I use it because I also had to rewrite the entire SearchHandler logic due to a huge number of bugs.

**Comment by arahmancsd** [comment]
Facing a similar issue with >= iPhone 8

**Comment by codingL3gend** [comment]
@kofanov i definitely agree with all of the bugs present in Maui. it's been annoying that my UI breaks with just about any update to the framework. i really want to champion this framework and recommend it but the issues have made that a bit challenging. having to put in too many workarounds just to keep the UI stable and across platforms is a chore

**Comment by ewerspej** [comment]
> @kofanov i definitely agree with all of the bugs present in Maui. it's been annoying that my UI breaks with just about any update to the framework. i really want to champion this framework and recommend it but the issues have made that a bit challenging. having to put in too many workarounds just to keep the UI stable and across platforms is a chore

That's the fate of early adopters with many technologies. Although, I do agree that MAUI seems to have been released prematurely. Then again, it's w highly competitive market. I think that Microsoft needs to extend the support for Xamarin.Forms until at least the end of 2024, because it will take some time to address the various issues of MAUI. 

However, as a community, we can all help make this framework better. The idea of MAUI is awesome and I'm already in love with all the cool new things, especially the single project and the .NET multi-targeting approach. So let's help make it better by reporting issues, finding workarounds and providing bugfixes.

**Comment by arahmancsd** [comment]
@sandydunlop I think I found a bit easy way to sort this. It is based on @kofanov solution.

Since I want this to happen on certain pages, so for a demo, I identified those pages from the title and created a dictionary to add the CustomTileView.

However, the rest looks the same. instead of OnApplyTemplate, I used the constructor. My `ContentPage` inherits from a `BaseContentPage`

CustomTileView
```
#if IOS
    protected override Size ArrangeOverride(Rect bounds)
    {
        // It is either zero or 20.x+, but I am ok with 25 here
        if (bounds.Height <= 25)
            bounds.Height = 44;

        return base.ArrangeOverride(bounds);
    }
#endif
```

and BaseContentPage constructor
`Shell.SetTitleView(this, new CustomTitleView());`

The result
![12](https://user-images.githubusercontent.com/7424078/234991899-ac8b99c1-079d-464d-86cf-8a183da5155f.png)

**Comment by codingL3gend** [comment]
@arahmancsd and it's always consistently there and doesn't disappear on navigations?

**Comment by arahmancsd** [comment]
> @arahmancsd and it's always consistently there and doesn't disappear on navigations?

This is consistent on defined pages, navigating back and forward doesn't have any effect on this issue on my part.  However, if you faced inconsistency, you can call `InvalidateMeasure()` on `OnNavigatedTo` or `OnAppearing`

**Comment by Dids** [comment]
How can issues like this still exist in MAUI? Issues like this absolutely plagued Xamarin.Forms, especially after the whole Shell mess, and sadly it looks like MAUI is still using Shell and still exhibiting the same inconsistency and bugs, with no official word on any fixes, ETAs or even acknowledgements.

It's become increasingly hard to sell XF/MAUI to my clients, and I'm not sure how much longer they can remain patient, while I hack my way around seemingly trivial issues such as this one.

**Comment by ewerspej** [comment]
For this particular issue there is a (temporary) workaround in the [MAUI Fixes and Workarounds NuGet package](https://www.nuget.org/packages/PureWeen.Maui.FixesAndWorkarounds) provided by Shane Neuville from the MAUI team. It's not ideal, but at least it works.

**Comment by ewerspej** [comment]
@rezamohamed Yes, that should be possible using a Grid, for example:

```xml
<Shell.TitleView>
  <Grid>
    <Label
      Text="Welcome to MAUI"
      HorizontalOptions="Center"
      VerticalOptions="Center" />
  </Grid>
</Shell.TitleView>
```

Generally, this type of question is suitable for Stack Overflow.

**Comment by rezamohamed** [comment]
> Generally, this type of question is suitable for Stack Overflow.

Typically, yes, I have checked SO as well. This appears to be a bug since it doesn't work with Flyouts, so I wanted to check here.

Even with a grid, the text doesn't center align correctly.

**Comment by ewerspej** [comment]
> Even with a grid, the text doesn't center align correctly.

It doesn't align or it doesn't align correctly? Does it align somewhat in the center or is the LayoutOption completely ignored? Which platform?

**Comment by leofalcao10** [comment]
I tried @ewerspej solution of using “ PureWeen.Maui.FixesAndWorkarounds”. I followed the tutorial on the link but it still doesn’t work on iOS. The top bar get a wrong position, but if I navigate to another page and click on go back button, the position of the top bar fixes.
![image](https://github.com/dotnet/maui/assets/54149744/879b70a5-6c15-46cf-bf7f-ef16350f863d)
![image](https://github.com/dotnet/maui/assets/54149744/d108d19a-9483-44d2-b651-4445c5763dea)
![image](https://github.com/dotnet/maui/assets/54149744/9b8b9fee-44ca-4353-99dd-53ee999e73a7)

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio Enterprise 17.9.0 Preview 2.1. Not repro on iOS platforms with sample project.
[maui-samples-main.zip](https://github.com/dotnet/maui/files/13826506/maui-samples-main.zip)
![Screenshot 2024-01-04 130825](https://github.com/dotnet/maui/assets/77434865/bd3242c6-9444-46b1-9cea-f83754eef0b3)
